### PR TITLE
Change fetch_transactions_for_block interface

### LINKED ISSUE #370: Change `fetch_transactions_for_block` into `prepare_transactions_with_result`
If we use `TxFIlter::fetch_transactions_for_block`, then proposers have to execute the same transactions twice.
This is because preparing transactions inherently requires the execution of transactions.
We should change the interface in order to avoid the double execution.

### PR DESCRIPTION
**Pull request by Byeongjee**
Resolves #370 
`mem_pool.top_transactions` should be removed.
We may need a separate interface dedicated to handle `ready_transactions` method.

### CODE REVIEW

**Comment by Byeongjee** [comment]
The first three commits deal with some minor issues on `mold`.
These changes should have been applied in #260. It was my mistake.

File: core/src/miner/mem_pool.rs [hunk 1]
```diff
@@ -216,33 +216,8 @@ impl MemPool {
...
-            chosen_transactions.iter().map(|tx_with_gas| tx_with_gas.tx_with_metadata.inserted_timestamp).max();
-
-        PendingTransactions {
-            transactions: chosen_transactions
-                .iter()
-                .map(|tx_with_gas| tx_with_gas.tx_with_metadata.tx.clone())
-                .collect(),
-            last_timestamp,
-        }
+    pub fn all_pending_transactions_with_metadata(&self) -> Vec<TransactionWithMetadata> {
```
Comment (Byeongjee) [inline core/src/miner/mem_pool.rs:219]: I'm worried about this costly clone.
The best solution would be creating a slice from `transaction_pool.pool`, but unfortunately, it was impossible.
I'm still looking for a nice solution.

Comment (Byeongjee) [reply core/src/miner/mem_pool.rs:219]: This is why I put **do_not_merge** label.
If you think this is fine, I can just merge it.
Otherwise, I'll continue looking for a solution.

File: coordinator/src/lib.rs [hunk 2]
```diff
@@ -57,6 +57,14 @@ impl BlockExecutor for Coordinator {
         unimplemented!()
     }
 
+    fn prepare_transactions_with_result<'a>(
```
Comment (sgkim126) [inline coordinator/src/lib.rs:60]: It seems that it doesn't return the result.

File: core/src/miner/mem_pool.rs (hunk 1 above)
Comment (sgkim126) [reply core/src/miner/mem_pool.rs:219]: I think it's okay to merge this commit with leaving the `XXX` comment if fixing it takes too much time.

File: coordinator/src/lib.rs (hunk 2 above)
Comment (Byeongjee) [reply coordinator/src/lib.rs:60]: That's right, but as you can see, I actually haven't implemented any of the methods in the file yet.
The implementation would be heavily related to some coordinator-side decisions, so I have to discuss with @dynaxis and other team members.

I think we should implement them soon so I'll have a discussion and work on it, but not in this PR.

**Comment by Byeongjee** [comment]
@sgkim126 I fixed the problem above by just simply returning an `impl Iterator`.
I think returning an iterator is not problematic, and it should be more efficient than the previous revision which returned a cloned vector.

File: coordinator/src/lib.rs (hunk 2 above)
Comment (sgkim126) [reply coordinator/src/lib.rs:60]: Even you haven't implemented it yet, you can declare the result type. If you cannot decide how to implement the result type yet, you should not name this method as `prepare_transactions_with_result`.

File: core/src/block.rs [hunk 3]
```diff
@@ -187,6 +187,15 @@ impl<'x> OpenBlock<'x> {
         Ok(())
     }
 
+    pub fn prepare_block_from_transactions(
+        &mut self,
+        block_executor: &dyn BlockExecutor,
+        transactions: &[&TransactionWithMetadata],
```
Comment (sgkim126) [inline core/src/block.rs:193]: How about making this function receives the `Iteration` instead of the slice of the reference??

File: coordinator/src/lib.rs (hunk 2 above)
Comment (Byeongjee) [reply coordinator/src/lib.rs:60]: Oh, I misunderstood your concern.
You're worrying because using the term `result` without returning an explicit `Result` type is confusing, right?
Actually I didn't mean this method to return a result type.
The method does two things.
1. Choose transactions to be included in the next block and return them
2. Execute the chosen transactions on the given `context`.

Previously Dohyung suggested the name for a reason in Discord.
Could you suggest a better name based on its behavior?
Maybe `prepare_block` works, because it's easy to infer that this method execute transactions (One of the argument is a mutable reference of `StorageAccess`)

https://discordapp.com/channels/682525519711043609/683931185982079019/699187814977372225
> @byeongjee renaming remove_transactions to filter_transactions seems a good idea. And your idea of splitting methods to open_block, close_block and the others also sounds great. But naming-wise, select_and_execute prefix seems very straight on what it does, but looks a bit verbose to me. And we don’t even select and then execute Txes. Maybe prepare_block would be better, IMHO. But I’m not sure. The exact meaning is ‘select Txes for inclusion in a block to propose by ordering and executing Txes, and then return the root hash’. So if we are to name it to the meaning, it would be more like prepare_transactions_with_result.

File: core/src/block.rs (hunk 3 above)
Comment (Byeongjee) [reply core/src/block.rs:193]: You mean `Iterator`, right?
I'll change the function.
I'm not sure if it's good to use the `Iterator` type in `TxFilter` and `BlockExecutor` methods (because the traits are high-level interfaces exposed to Coordinator) so I didn't change them yet.
Please let me know if you think it's better to apply the change.

File: coordinator/src/lib.rs (hunk 2 above)
Comment (Byeongjee) [reply coordinator/src/lib.rs:60]: I used `prepare_block` for now.

File: core/src/block.rs (hunk 3 above)
Comment (sgkim126) [reply core/src/block.rs]: Because the `Iterator` is a more abstracted type than the slice, I couldn't understand your concerns.

Comment (sgkim126) [reply core/src/block.rs]: Did you mean it should be serializable?

File: core/src/block.rs [hunk 4]
```diff
@@ -187,6 +188,16 @@ impl<'x> OpenBlock<'x> {
         Ok(())
     }
 
+    pub fn prepare_block_from_transactions<'a>(
+        &mut self,
+        block_executor: &dyn BlockExecutor,
+        transactions: impl Iterator<Item = &'a TransactionWithMetadata>,
+    ) {
+        let transactions = transactions.collect::<Vec<_>>();
```
Comment (sgkim126) [inline core/src/block.rs:196]: If you change `prepare_block` to receive the iterator also, you don't need to create the vector in here.

File: core/src/block.rs (hunk 3 above)
Comment (Byeongjee) [reply core/src/block.rs]: I was slightly worried because the coordinator is directly tied to FML system, and it seemed that it wouldn't be easy to manipulate `Iterator` types in the FML system.

However, as a second thought, if we assume that the coordinator and the host are built and run together in the same process, it wouldn't be problematic to pass `Iterator` type from the host to the coordinator, and it will just be the coordinator's task to unpack the iterator.

In summary, never mind, I'll use iterator type (also in `filter_transactions` method)

File: core/src/block.rs (hunk 4 above)
Comment (Byeongjee) [reply core/src/block.rs:196]: That's true but we still need something complex here, because we have to copy or reference the given iterator somehow because of lifetime parameters of `prepare_block` require it.

Comment (sgkim126) [reply core/src/block.rs]: You don't need to pass the `Iterator` as `&mut`, `Box<dyn Iterator<Item = &'a TransactionWithMetadata> + 'a>` is enough.
If you think it's complex yet, you can make the `'a` and the Iterator parameters of `BlockExecutor`, like 
```rust
pub trait BlockExecutor<'a, Transactions>: Send + Sync  where Transactions: Iterator<Item = &'a TransactionWithMetadata>
```
But it will make the other code that uses the trait more complex.

Comment (Byeongjee) [reply core/src/block.rs]: Oh, I didn't think about using a lifetime keyword in the `Box` type. Thanks, I'm still getting used to it.
I actually thought about the latter option, but then I couldn't instantiate `dyn BlockExecutor` object.

Comment (Byeongjee) [reply core/src/block.rs]: I just realized that I could just use `&mut dyn Iterator` as well.
But I think `Box<dyn Iterator>` is better than `&mut dyn Iterator`, since it states about the reusability more clearly.
I'll use `Box<dyn Iterator>` for both `prepare_block` and `filter_transactions`.

**Comment by Byeongjee** [comment]
@sgkim126 I changed the argument into `Box<dyn Iterator>` type as you advised.
I changed `filter_transactions` method, too, for consistency.
